etw/native/build.gradle

```diff
@@ -0,0 +1,135 @@
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+plugins {
+    id 'cpp-library'
+    id 'cpp-unit-test'
+    id 'visual-studio'
+}
+
+// TODO honor skipWinNative property
+
+def javaIncludeDir = new File(System.env.'JAVA_HOME', "include")
+def programFilesX86 = System.env.'ProgramFiles(x86)' ?: 'C:/Program Files (x86)'
+
+// TraceLoggingProvider.h in Windows 10 SDK
+def winSdkDir = "$programFilesX86/Windows Kits/10"
+if (System.env.APPINSIGHTS_WIN_SDK_PATH) {
+    winSdkDir = System.env.APPINSIGHTS_WIN_SDK_PATH
+    logger.info "Windows SDK path set with environment variable, APPINSIGHTS_WIN_SDK_PATH"
+}
+def vsToolsDir = "$programFilesX86/Microsoft Visual Studio 14.0"
+if (System.env.APPINSIGHTS_VS_PATH) {
+    vsToolsDir = System.env.APPINSIGHTS_VS_PATH
+    logger.info "Visual Studio Build Tools path set with environment variable, APPINSIGHTS_VS_PATH"
+}
+
+def winSdkLibPath
+if (System.env.APPINSIGHTS_WIN_SDK_LIB_PATH) {
+    winSdkLibPath = System.env.APPINSIGHTS_WIN_SDK_LIB_PATH
+    logger.info "Windows SDK Lib path set with environemnt variable, APPINSIGHTS_WIN_SDK_LIB_PATH=${System.env.APPINSIGHTS_WIN_SDK_LIB_PATH}"
+} else {
+    winSdkLibPath = "$winSdkDir/Lib/10.0.18362.0/um"
+}
+
+logger.info "Windows SDK:  $winSdkDir"
+logger.info "SDK Lib path: $winSdkLibPath"
+logger.info "Visual Studio Build Tools: $vsToolsDir"
+
+toolChains.withType(VisualCpp).configureEach  {
+    // FIXME can this toolchain be addressed directly?
+    it.installDir vsToolsDir
+    it.windowsSdkDir winSdkDir
+}
+
+ext {
+    headersDir = "$projectDir/src/main/headers"
+}
+library {
+    baseName = "applicationinsights-java-etw-provider"
+    linkage = [ Linkage.SHARED ]
+    targetMachines = [ machines.windows.x86, machines.windows.x86_64 ]
+    privateHeaders.from file(headersDir)
+    dependencies {
+        implementation files(file("$winSdkLibPath/x64/advapi32.lib"))
+        implementation files(file("$winSdkLibPath/x86/advapi32.lib"))
+    }
+}
+
+
+logger.info "$NATIVE_BUILD_VARIANT_PROPERTY = ${project.properties[NATIVE_BUILD_VARIANT_PROPERTY]}"
+logger.info "$NATIVE_VERBOSE_OUTPUT_PROPERTY = ${project.properties[NATIVE_VERBOSE_OUTPUT_PROPERTY]}"
+
+// NOTE: these options apply to Visual Studio Build Tools (cl.exe)
+tasks.withType(CppCompile).configureEach {
+    includes file("${System.env.JAVA_HOME}/include")
+    includes file("${System.env.JAVA_HOME}/include/win32")
+
+    compilerArgs.add '/W4'
+    compilerArgs.add '/std:c++14'
+    compilerArgs.add '/EHsc'
+    compilerArgs.add '/sdl'
+
+    if (project.hasProperty(NATIVE_BUILD_VARIANT_PROPERTY) && 'release'.equalsIgnoreCase(project.properties[NATIVE_BUILD_VARIANT_PROPERTY])) {
+        logger.info "Configuring ${it} for release build"

```

Comment:
That's cool........ you can also use .vcxproj and **msbuild** to prescribe default flags for Release and Debug, without manually passing individual compiler args and flags. It'd be more natural for C++ dev dealing with Visual Studio to just edit the .vcxproj file for that.

Reply:
I did try that, but it was a bit difficult working with it and the gradle plugin. I'm more familiar with gradle than vcxproj :/ You can still generate the vcxproj with `.\gradlew :etw:native:visualStudio` if one prefers to work in Visual Studio.